transfers/dispatcher.go

```diff
@@ -0,0 +1,335 @@
+// whether the files to be transferred require staging.
+//
+// A client can request the status of ongoing and completed transfers, which the dispatcher fetches
+// from the store. A client can also request that an ongoing transfer be deleted, which the
+// dispatcher propagates to the store, stager, mover, and manifestor.
+//
+// When the dispatcher is stopped, it stops the store, stager, mover, and manifestor.
+
+// dispatcher global state
+var dispatcher dispatcherState
+
+type dispatcherState struct {
+	Channels dispatcherChannels
+}
+
+type dispatcherChannels struct {
+	RequestTransfer  chan Specification // used by client to create a new transfer
+	ReturnTransferId chan uuid.UUID     // returns task ID to client
+
+	CancelTransfer chan uuid.UUID // used by client to cancel a transfer
+
+	RequestStatus chan uuid.UUID      // used by client to request transfer status
+	ReturnStatus  chan TransferStatus // returns task status to client
+
+	Error chan error    // internal -> client error propagation
+	Stop  chan struct{} // used by client to stop task management
+}
+
+func newDispatcherChannels(maxConnections int) dispatcherChannels {
+	return dispatcherChannels{
+		RequestTransfer:  make(chan Specification, maxConnections),
+		ReturnTransferId: make(chan uuid.UUID),
+		CancelTransfer:   make(chan uuid.UUID, maxConnections),
+		RequestStatus:    make(chan uuid.UUID, maxConnections),
+		ReturnStatus:     make(chan TransferStatus),
+		Error:            make(chan error),
+		Stop:             make(chan struct{}),
+	}
+}
+
+func (channels *dispatcherChannels) Close() {
+	close(channels.RequestTransfer)
+	close(channels.ReturnTransferId)
+	close(channels.CancelTransfer)
+	close(channels.RequestStatus)
+	close(channels.ReturnStatus)
+	close(channels.Error)
+	close(channels.Stop)
+}
+
+func (d *dispatcherState) Start() error {
+	d.Channels = newDispatcherChannels(config.Service.MaxConnections)
+	go d.process()
+	return <-d.Channels.Error
+}
+
+func (d *dispatcherState) Stop() error {
+	d.Channels.Stop <- struct{}{}
+	err := <-d.Channels.Error
+	d.Channels.Close()
+	return err
+}
+
+func (d *dispatcherState) CreateTransfer(spec Specification) (uuid.UUID, error) {
+	d.Channels.RequestTransfer <- spec
+	select {
+	case id := <-d.Channels.ReturnTransferId:
+		return id, nil
+	case err := <-d.Channels.Error:
+		return uuid.UUID{}, err
+	}
+}
+
+func (d *dispatcherState) GetTransferStatus(transferId uuid.UUID) (TransferStatus, error) {
+	d.Channels.RequestStatus <- transferId
+	select {
+	case status := <-d.Channels.ReturnStatus:
+		return status, nil
+	case err := <-d.Channels.Error:
+		return TransferStatus{}, err
+	}
+}
+
+func (d *dispatcherState) CancelTransfer(transferId uuid.UUID) error {
+	d.Channels.CancelTransfer <- transferId
+	err := <-d.Channels.Error
+	if err != nil {
+		slog.Error(fmt.Sprintf("Transfer %s: %s", transferId.String(), err.Error()))
```

Comment:
Not a big deal, but if we log cancel errors, should we also log errors for other dispatcher functions?

Reply:
Good idea. Though the reason I do this here is that cancellation errors aren't propagated, so logging them is the only way to capture them. I think at length we should audit error handling and logging (and other things too) to make sure they're being done sensibly.